From 80e49892a807387a2c15f351d372134fa738fe8f Mon Sep 17 00:00:00 2001
From: Isaku Yamahata <isaku.yamahata@intel.com>
Date: Tue, 8 Mar 2022 11:06:01 -0800
Subject: [PATCH 51/99] [REVERTME] i386/tdx: Make GetQuote handler accept old
 ABI

TDG.VP.VMCALL<GetQuote> in GHCI1.0 spec is changed so that the shared GPA
size is passed in R13.  With The old GHCI R13 isn't used and the buffer size
is assumed to be fixed 8KBytes.

If R13 input value is zero, it assumes the old GetQuote ABI and handle it
for compatibility.

Signed-off-by: Isaku Yamahata <isaku.yamahata@intel.com>
---
 target/i386/kvm/tdx.c | 22 +++++++++++++++++++---
 1 file changed, 19 insertions(+), 3 deletions(-)

diff --git a/target/i386/kvm/tdx.c b/target/i386/kvm/tdx.c
index b3412280cb..2e558d165b 100644
--- a/target/i386/kvm/tdx.c
+++ b/target/i386/kvm/tdx.c
@@ -901,7 +901,14 @@ static void tdx_handle_get_quote(X86CPU *cpu, struct kvm_tdx_vmcall *vmcall)
         return;
     }
     if (buf_len == 0) {
-        return;
+        /*
+         * REVERTME: Accept old GHCI GetQuote with R13 buf_len = 0.
+         * buf size is 8KB. also hdr.out_len includes the header size.
+         */
+#define GHCI_GET_QUOTE_BUFSIZE_OLD      (8 * 1024)
+        warn_report("Guest attestation driver uses old GetQuote ABI.(R13 == 0) "
+                    "Please upgrade guest kernel.\n");
+        buf_len = GHCI_GET_QUOTE_BUFSIZE_OLD;
     }
 
     if (address_space_read(&address_space_memory, gpa, MEMTXATTRS_UNSPECIFIED,
@@ -916,10 +923,19 @@ static void tdx_handle_get_quote(X86CPU *cpu, struct kvm_tdx_vmcall *vmcall)
      * leak.  Enforce it.  The initial value of them doesn't matter for qemu to
      * process the request.
      */
-    if (le64_to_cpu(hdr.error_code) != TDX_VP_GET_QUOTE_SUCCESS ||
-        le32_to_cpu(hdr.out_len) != 0) {
+    if (le64_to_cpu(hdr.error_code) != TDX_VP_GET_QUOTE_SUCCESS
+        /* || le32_to_cpu(hdr.out_len) != 0 */) {
         return;
     }
+    if (le32_to_cpu(hdr.out_len) > 0) {
+        /* REVERTME: old shared page format. */
+        warn_report("Guest attestation driver or R3AAL uses old GetQuote format."
+                    "(out_len > 0) Please upgrade driver or R3AAL library.\n");
+        if (le32_to_cpu(hdr.out_len) + sizeof(hdr) > buf_len) {
+            return;
+        }
+        hdr.out_len = cpu_to_le32(0);
+    }
 
     /* Only safe-guard check to avoid too large buffer size. */
     if (buf_len > TDX_GET_QUOTE_MAX_BUF_LEN ||
-- 
2.31.1

